pkg/controller/atlasproject/encryption_at_rest.go

```diff
@@ -34,6 +45,116 @@ func ensureEncryptionAtRest(ctx *workflow.Context, projectID string, project *md
 	return workflow.OK()
 }
 
+func readEncryptionAtRestSecrets(kubeClient client.Client, service *workflow.Context, encRest *mdbv1.EncryptionAtRest, parentNs string) error {
+	if encRest == nil {
+		return nil
+	}
+
+	if encRest.AwsKms.Enabled != nil && *encRest.AwsKms.Enabled && encRest.AwsKms.SecretRef.Name != "" {
+		watchObj, err := readAndFillAWSSecret(kubeClient, parentNs, &encRest.AwsKms)
+		service.AddResourcesToWatch(*watchObj)
+		if err != nil {
+			return err
+		}
+	}
+
+	if encRest.GoogleCloudKms.Enabled != nil && *encRest.GoogleCloudKms.Enabled && encRest.GoogleCloudKms.SecretRef.Name != "" {
+		watchObj, err := readAndFillGoogleSecret(kubeClient, parentNs, &encRest.GoogleCloudKms)
+		service.AddResourcesToWatch(*watchObj)
+		if err != nil {
+			return err
+		}
+	}
+
+	if encRest.AzureKeyVault.Enabled != nil && *encRest.AzureKeyVault.Enabled && encRest.AzureKeyVault.SecretRef.Name != "" {
+		watchObj, err := readAndFillAzureSecret(kubeClient, parentNs, &encRest.AzureKeyVault)
+		service.AddResourcesToWatch(*watchObj)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func readAndFillAWSSecret(kubeClient client.Client, parentNs string, awsKms *mdbv1.AwsKms) (*watch.WatchedObject, error) {
+	fieldData, watchObj, err := readSecretData(kubeClient, awsKms.SecretRef, parentNs, "CustomerMasterKeyID", "Region", "RoleID")
+	if err != nil {
+		return watchObj, err
+	}
+
+	fillStructFields(fieldData, awsKms)
+	return watchObj, nil
+}
+
+func readAndFillGoogleSecret(kubeClient client.Client, parentNs string, gkms *mdbv1.GoogleCloudKms) (*watch.WatchedObject, error) {
+	fieldData, watchObj, err := readSecretData(kubeClient, gkms.SecretRef, parentNs, "ServiceAccountKey", "KeyVersionResourceID")
+	if err != nil {
+		return watchObj, err
+	}
+	fillStructFields(fieldData, gkms)
+	return watchObj, err
+}
```

Comment:
My 2 cents, why not make an explicit static conversion here from the secret data fields to the cloud specific struct?
I think it will be less code, and more readable than having to implement those tricky `readSecretData` and `fillStructFields`. It would look not very different from the `ToAtlas` conversions above, which are pretty straightforward.

Why was this way preferred?

Reply:
I removed the reflection and added an explicit filling for cloud provider structures